rbac/migration_tool/in_memory_tuples.py

```diff
@@ -25,15 +25,202 @@ class RelationTuple(NamedTuple):
+
+        # Iterate over each group
+        for key, group_tuples in grouped_tuples.items():
+            remaining_tuples = set(group_tuples)
+            remaining_predicates = list(predicates) if match_once else predicates
+            i = 0
+            matching_tuples = set()
+            success = True
+
+            # Attempt to match all predicates within the group
+            # Using each predicate only once if requested
+            while remaining_predicates and i < len(remaining_predicates):
+                predicate = remaining_predicates[i]
+                found = False
+                for rel in remaining_tuples:
+                    if predicate(rel):
+                        matching_tuples.add(rel)
+                        remaining_tuples.remove(rel)
+                        found = True
+                        break  # Move to next predicate
+                if not found:
+                    success = False
+                    break  # Predicate not satisfied in this group
+                if match_once:
+                    remaining_predicates.pop(i)
+                else:
+                    i += 1
+
+            if require_full_match and remaining_tuples or not success:
+                unmatched_groups[key] = TupleSet(self._full_set, group_tuples)
+                continue
+
+            matching_groups[key] = TupleSet(self._full_set, matching_tuples)
+
+        return matching_groups, unmatched_groups
+
+    def traverse_subject(
+        self,
+        predicates: List[RelationPredicate],
+        require_full_match: bool = True,
+        match_once: bool = True,
+    ) -> "TupleSet":
+        """
+        Traverse through the tuples to find and match tuples where the subject is the resource of the current tuple.
+
+        The returned set contains all tuples that match the predicates
+        and have a resource that is the subject of a tuple in this set.
+
+        Args:
+            predicates (List[RelationPredicate]):
+                A list of predicate functions that take a RelationTuple
+                and return a boolean indicating if the tuple matches the condition.
+            require_full_match (bool, optional):
+                If True, each predicate must be matched at least once. Defaults to True.
+            match_once (bool, optional):
+                If True, each tuple in the traversed list should be matched by at least one predicate. Defaults to True.
+        Returns:
+            TupleSet: A set of matched RelationTuples.
+        """
+        # For each tuple, find all the tuples where the subject is the resource
+        # of the current tuple
+        matched: set[RelationTuple] = set()
+
+        for tuple in self:
+            # Intentionally use full set here to traverse to tuples which may not be in this set
+            traversed = self._full_set.find_tuples(
+                resource(tuple.subject_type_namespace, tuple.subject_type_name, tuple.subject_id)
```

Comment:
The key bit here is that this uses the full set of tuples to search, not just what was filtered in this result set. This result set is just used as the base for traversal (using the `resource` predicate for each subject in this set).